lib/improver/nbhood.py

```diff
@@ -246,27 +246,102 @@ def _apply_kernel_for_smoothing(self, cube, ranges):
+            open_grid_summed_squared = np.sum(open_grid**2.).astype(float)
             kernel[:] = (
-                (np.cumprod(ranges)[-1] - np.sum([x**2. for x in open_grid])) /
-                np.cumprod(ranges)[-1]
-            )
+                (np.prod(ranges) - open_grid_summed_squared) / np.prod(ranges))
             mask = kernel < 0.
         kernel[mask] = 0.
         # Smooth the data by applying the kernel.
         cube.data = scipy.ndimage.filters.correlate(
             data, kernel, mode='nearest') / np.sum(kernel)
         return cube
 
+
+class NeighbourhoodProcessing(object):
+    """
+    Apply a neigbourhood processing kernel to a thresholded cube.
+
+    When applied to a thresholded probabilistic cube, it acts like a
+    low-pass filter which reduces noisiness in the probabilities.
+
+    The kernel will presently only work with projections in which the
+    x grid point spacing and y grid point spacing are constant over the
+      entire domain, such as the UK national grid projection
+
+    A maximum kernel radius of 500 grid cells is imposed in order to
+    avoid computational ineffiency and possible memory errors.
+
+    """
+
+    # Max extent of kernel in grid cells.
+    MAX_KERNEL_CELL_RADIUS = 500
+
+    def __init__(self, kernel_method, radii_in_km, lead_times=None,
+                 unweighted_mode=False):
+        """
+        Create a neighbourhood processing plugin that applies a smoothing
+        kernel to points in a cube.
+
+        Parameters
+        ----------
+
+        kernel_method : str
+            Name of the neighbourhood kernel to use. Options: 'circular'.
+        radii_in_km : float or List (if defining lead times)
+            The radii in kilometres of the neighbourhood kernel to
+            apply. Rounded up to convert into integer number of grid
+            points east and north, based on the characteristic spacing
+            at the zero indices of the cube projection-x/y coords.
+        lead_times : None or List
+            List of lead times or forecast periods, at which thel radii
+            within radii_in_km are defined. The lead times are expected
+            in hours.
+        unweighted_mode : boolean
+            If True, use a circle with constant weighting.
+            If False, use a circle for neighbourhood kernel with
+            weighting decreasing with radius.
+
+        """
+        if (hasattr(Kernels, kernel_method) and
+                callable(getattr(Kernels, kernel_method))):
+            self.kernel_method = getattr(Kernels, kernel_method)
+        else:
+            methods = [method for method in dir(Kernels)
+                       if callable(getattr(Kernels, method))]
+            msg = ("The requested kernel method: {} does not exist within the "
+                   "available options for the kernel: {}".format(
+                       kernel_method, methods))
+            raise AttributeError(msg)
```

Comment:
This code section for testing whether the Kernels class contains a callable object `kernel_method` should be in the Kernels class.